featurebyte/persistent/mongo.py

```diff
@@ -178,11 +179,48 @@ async def _find(
         tuple[Iterable[Document], int]
             Retrieved documents and total count
         """
-        cursor = self._db[collection_name].find(
-            filter=query_filter, projection=projection, session=self._session
+        cursor: AgnosticCursor = cast(
+            AgnosticCursor,
+            await self._get_iterator(collection_name, query_filter, projection, sort_by),
         )
         total = await self._db[collection_name].count_documents(query_filter, session=self._session)
 
+        if page_size > 0:
+            skips = page_size * (page - 1)
+            cursor = cursor.skip(skips).limit(page_size)
+
+        result: list[Document] = await cursor.to_list(total)
+        return result, total
+
+    async def _get_iterator(
+        self,
+        collection_name: str,
+        query_filter: QueryFilter,
+        projection: Optional[dict[str, Any]] = None,
+        sort_by: Optional[list[tuple[str, SortDir]]] = None,
+    ) -> AsyncIterator[Document]:
```

Comment:
Is the return type annotation accurate? There is quite a few manual casting related to this.

Reply:
The casting is needed due to Document (MutableMapping[str, Any]) being a more general type vs Dict[str, Any]. We can update the persistent type to be more specific to avoid the casting

Reply:
the casting to AgnosticCursor is required since `_get_iterator` has a more generic return type but we needed to perform actions on the returned object being aware that it's a cursor